### ISSUE TITLE
[Question] Way to conbine ssim loss and l1 loss

### ISSUE THREAD

**Issue by ljijjiang1234** [mitsuba-renderer/mitsuba3#934]
## Summary

Hi , I am working on  autodiff for some parameters estimation , I want improve ssim of final result , so I want to add ssim loss into the code , Unfortunately , all tutorial was working on l1 loss for inverse rendering , so I try to use dr.warp_ad to convert TensorXf into pytorch tensor and use pytorch_ssim loss and l1 loss add them together and then return as an dr.tensorXF , however it not work as I expected seems like the  gradiant was lost in torch loss function calculation.

## Description

First I import the package https://github.com/Po-Hsun-Su/pytorch-ssim , and use it function
Second I define in dr to torch interation funtion as following :
~~~python
ssim_fun = pytorch_ssim.SSIM().cuda()
l1_fun = torch.nn.L1Loss()
@dr.wrap_ad(source='drjit', target='torch')
def loss(img1,img2,ssim_weight=0.5, l1_weight=0.5):
    img1=img1.cuda()
    img2=img2.cuda()
    image1 = img1.unsqueeze(0)
    image2 = img2.unsqueeze(0)
    # Compute the SSIM loss
    SSIM_loss=1 * (1 - ssim_fun(image1,image2))
    # Compute the L1 loss
    l1_loss = 1*l1_fun(image1,image2)
    # Combine the losses
    loss = SSIM_loss+l1_loss
    return loss
~~~

Finally I try to use it in inverse loop to do inverse as your tutorial:
~~~python
for it in range(320):
    img = mi.render(scene_uncom, params_uncom, seed=it, spp=12)
    image=torch_func(ccm,img)
    dr.enable_grad(image)
    dr.enable_grad(img_ref)
    loss = dr.cuda.ad.Float(loss(image,img_ref).torch())
    dr.backward(loss)
    opt.step()
    loss_hist.append(loss)
    params_uncom.update(opt)
    print(f"Iteration {it:02d}: error={loss[0]:6f}")
~~~

and I get the bug as following:
<img width="1212" alt="image" src="https://github.com/mitsuba-renderer/mitsuba3/assets/73884949/ee0f1108-e152-4757-acae-c9f881240331">

## Conclusion:
Would you kindly like to share with me some idea to calculate ssim loss and use it to do inverse rendering rendering process in drjit cuda backend?

**Comment by njroussel** [comment]
Hi @ljijjiang1234 

I think this should be possible

As far as I can tell this line is wrong
```python
loss = dr.cuda.ad.Float(loss(image,img_ref).torch())
```
This is overly complicated. The wrapped (`wrap_ad`) `loss` function does a computation in PyTorch but returns a Dr.Jit type. In this line you ar converting it to PyTorch again, just cast it back to Dr.Jit. You should just have:
```python
loss_ = loss(image, img_ref)
```

I'm guessing your snippets are just pseudo-code. If not, be careful with this shadowing of the `loss` function, sometimes it's a function, sometimes it's a variable...

I also think neither of these lines should be necessary:
```python
dr.enable_grad(image)
dr.enable_grad(img_ref)
```
At the very leas it should not be needed for `img_ref`. I'm assuming that `torch_func` is another wrapped function? If it is, then you indeed shouldn't need it.

**Comment by ljijjiang1234** [comment]
Thanks for you patient answer, I have solved the question , just because I dont distinguish loss and loss_ you write , after that , use your advice code , can do auto diff with ssim loss. Finally this issue can be opened as a disscusstion for anyone else who want to use ssim_loss in mitsuba3 would be helpful.
